notes/conds.md

```diff
@@ -0,0 +1,83 @@
+| ------------ | --------------------------------- | ------------------------------------------ | ---------------------------------------------- |
+| `/${string}` |`[] -> Any`                        | Selects value using JSON Pointer            | `["/foo/0"]`                                  |
+|              |`[] -> Any`                        | Selects whole value like "" in JSON Pointer | `[]`                                          |
+| `==`         |`[Exp<Any>, Exp<Any>] -> Bool`     | Equality                                    | `["==", ["/size"], 1]`                        |
+| `<`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than                                   | `["<",  ["/size"], 1024]`                     |
+| `<=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than or equal                          | `["<=", ["/size"], 1024]`                     |
+| `>`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than                                | `[">", ["@", "/size"], 0]`                    |
+| `>=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than or equal                       | `[">=", ["@", "/size"], 32]`                  |
+| `not`        |`[Expr<Bool>] -> Bool`             | Negation                                    | `["not", ["<=", ["/size"], 128]`              |
+| `some`       |`[Expr<Any[]>, Expr<Bool>] -> Bool`| Collection predicate                        | `["some", ["/to"], ["==", [], "hi@web.mail"]]`|
+| `match`      |`[Expr<Str>, Exp<Str>] -> Bool`    | String wildcard                             | `["match", ["/to"], "*@foo.com"]`             |
+| `or`         |`Exp<Bool>[] -> Bool`              | Logical **or** combinator                   | `["or", [">", ["/n"], 0], ["==", ["/n"], 0]]` |
+| `and`        |`Exp<Bool>[] -> Bool`              | Logical **and** combinator                  | `["and", [">", ["/n"], 0], ["<", ["/n"], 10]]`|
+
+## Examples
+
+Delegates `email/send` command on `did:key:alice` with a following constraint: Email is not send to address on `gmail.com`, or someone from `fission.codes` should be cc-ed.
+
+```js
+{
+  "sub": "did:key:alice",
+  "iss": "did:key:alice",
+  "aud": "did:key:bob",
+  "cmd": "email/send",
+  "cond": [
+    ["or",
+      // no recipient should be at `@gmail.com`
+      ["every", ["/to"], ["not", ["match", [], "*@gmail.com"]]],
+      // or cc should include @fission.codes
+      ["some", ["/cc"], ["match", [], "*@fission.codes"]]
+    ]
+  ]
+}
+```
+
+## logic programming language
+
+We are also considering logic programming alternative inspired by datomic query styntax.
+
+### concrete proposal for the inital release
+
+Here's a sketch that shows the features
+
+```js
+{
+  "iss": "did:key:alice",
+  // ...
+  "policy": [
+     ["$ARGS", "foo..bar.[].baz", "?x"],
+     [["some", "match"], "?x", "*@example.com"],
+      
+     ["$ARGS", "foo.quux", "?y"],
+     ["?y", ">", 42],
+     ["?x", "<", "?y"]
+  ]
+}
+```
+
```

Comment:
I put some more thought into this and sketched this in TS

```ts
type Clause =
  | Match
  // Comparison Operators
  | GreaterThan
  | GreaterOrEqualThan
  | LessThan
  | LessOrEqualThan
  | Equal
  | Like
  // Logic Operators
  | And
  | Or

/**
 * Selector starts with `.` character making it distinguishable from
 * all other operators.
 */
type Selector = `.{string}`

type Match =
  // Top level selector.
  | [selector: Selector, term: Term]
  // Nested Selector from the given entity
  | [selector: Selector, entity: Term, term: Term]


type GreaterThan = [">", Term, Term]
type GreaterOrEqualThan = [">=", Term, Term]
type LessThan = ["<", Term, Term]
type LessOrEqualThan = ["<=", Term, Term]
type Equal = ["==", Term, Term]
type And = ["and", ...Clause[]]
type Or = ["or", ...Clause[]]
type Like = ["match", Term<Str>, Term<Pattern>]


type Term<T extends Constant = Constant> = Variable<T> | T

/**
 * Variables start with `?` mark character
 */
type Variable<T extends Constant = Constant> = `?{string}` & { phantom?: T }


type Constant =
  | Str
  | number
  | boolean
  | null
  | Bytes

/**
 * Our strings are distinct from variables.
 */
type Str = Exclude<string, Variable>

/**
 * Pattern is a string that can contain wildcards.
 */
type Pattern = Exclude<string, Variable>

/**
 * Adopt bytes representation from DAG-JSON
 */
type Bytes = { "/": { "bytes": string }}
```

And according to the above types we could rewrite above example as follows

```js
```js
{
  "iss": "did:key:alice",
  // ...
  "policy": [
     [".foo..bar.[].baz", "?x"],
     ["match", "?x", "*@example.com"],
     [".foo.quux", "?y"],
     [">", "?y", 42],
     ["<", "?x", "?y"]
  ]
}
```

Not sure if this is better, but at least `or` / `and` combinators would fit better I think.

Reply:
In the future if we decide we want to add `exp` we could expose it via operator e.g.

```ts
["exp", "?deadline"]
```

Reply:
> [".foo.quux", "?y"],

Something I liked about the previous triple version is that it gave you in/out ports:

```js
 "policy": [
     ["$args", "foo..bar.[].baz", "?x"], // Start selecting
     ["match", "?x", "*@example.com"], // Filter
     ["?x", "quux.[3]", 42] // Continue to equality match
  ]
```

Which lets you express lots of things.

Is the idea that when you omit the first variable, it's assumed to act on `args`? I can work with that, though having form consistency is also a nice property

Reply:
> but at least or / and combinators would fit better I think.

I'm not sure that I follow from the example. Could you expand on that?

Reply:
Oh! You just mean this:

```
type And = ["and", ...Clause[]]
type Or = ["or", ...Clause[]]
```

Yeah that form makes sense 👍

Reply:
> Adopt bytes representation from DAG-JSON

yes agreed 👍

Reply:
One other thing that has been on my mind is that I think selector should somehow signal cardinality, perhaps `?x[]` because by default it would assume produce separate matches (at least implementations I'm familiar with). 

That would also remove need for `every` / `some` because `[match, "?x", "*@web.mail"]` would imply `some` and `[match, "?x[]", "*@web.mail"]` would imply `every`.

Maybe there is a strategy that can defer that decision until encountering `every` modifier  but at the moment I have difficulty seeing how it will work exactly.


Separately it seems neat to de able to control whether matches should aggregate or not.

Reply:
I'm also in favour of having an implicit `every` lifted for the "top" logic variables (ones at the top of the dependency graph). You turn that behaviour off with a `some`. I _think_ that's the most intuitive for someone reading/writing these.

Reply:
Also I liked your suggestion earlier that saying "the `foo` field must contain `bar`" would work for both of these cases:

```
{foo: "bar"}
// and
{foo: ["foo"", "bar", "baz]
```

...because it does contain "bar".

But I _think_ maybe I see what you meant before: it's not that we can't match on either, but you might want to say "this must not contain an array, only match on strings/numbers/bools/bytes/cids". It a good point. Let me think on it a bit more 🤔

We can always have a `cardinality` operator to make it compositional (because again: unification makes this fairly natural), but that feels like cheating 😁

Reply:
Wait, back up: don't we already get cardinality from the path selector?

```js
// Matching against: {"foo": {"bar": [1,2,3]}}

policy: [
  ["$", ".foo.bar", "?x"],
  ["==", "?x", [1, 2, 3]], // Single cardinality containing an array
  // the rest is redundant in this context, but for example:
  ["?x", ".[]", "?y"], // Bind all of x's elements to y
  ["some", "?y"],
  ["==", "?y", 2] // matches the `2` in the array
]
```

Does that give us what we need without special operators, or have I missed something? 🤔

Reply:
What I was trying to say if I have input like `{"foo": {"bar": [1,2,3]}}` and query like

```js
{
  policy: [
    ["?", ".foo.bar", "?x],
    [">", "?x", 1]
  ]
}
```

I would get following output frames

```js
{ x: 2 }
{ x: 3 }
```

Which is to say behavior is of a `some` and not of `every`. I think there's a need for a aggregation of sorts. Maybe I'm missing something.

Reply:
> ```js
> ["some", "?x"] // <-- unify correctly as a "some" matcher, no magic (see below)
> ["$args", ".foo..bar.[].baz", "?x"]
> ["match", "?x", "*@example.com"]
> ```

This is confusing to me, is it does not map to notation used by datomic. What does `[some ?x]` imply that `x` got bound ? It was bound because it is present in other clause.

If `?x` selects a list (which I think is the goal here) that match should be element of the `?x` and more specifically you may want to say that every element MUST match or some elements MUST match.

Lets ignore the first clause for a moment and assume that `x` got produced a substitution

```js
{ x: ["alice@example.com", "bob@web.mail"] }
```

Now executing `match` rule must either unify and forward above substitution, or fail. Which is to say that match must either:

1. Operate under some semantics and unify if any candidate matches pattern.
2. Operate under every semantics  and unify if all candidates match pattern.
3. Operate under cardinality 1 and expect a string candidate and fail otherwise.

I would expect it to be either 1 or 3. If we also need `match/every` or some other form of triggering option 2. If we say `match` operates under 3 than we also need `match/some` or some other form of triggering option 1.

Assuming above does make sense we need something along this lines instead, but now syntax is no longer intuitive and we started nesting things.

```js
[
  ["?", ".foo..bar.[].baz", "?x"]
  [["match", "?x", "*@example.com"], "?matches"]
  ["every", "?matches"]
]
```

We could probably remove the `?matches` and inline 2nd clause into 3rd, but then again I would personally find this far more readable and intuitive 

```js
[
  ["?", ".foo..bar.[].baz", "?x"]
  ["match/every", "?x", "*@example.com"]
]
```

Perhaps there is another more composable way express that you want to apply operator on members of the variable but it is not obvious to me.

Reply:
Here is even worse example, that also illustrates why I was in more favor of JSON pointer. Let's assume this is our data

```js
{
   "messages": [
      { "to": ["alice@web.mail", "malory@web.mail"], "body": "hello alice" },
      { "to": ["bob@gmail.com", "alice@web.mail"], "body": "hi bob" }
   }
}
```

What should be the `?to` bound to in this query ? Array of all emails addresses or each email address ?

```js
[
  ["?", ".messages[].to[]", "?to"]
]
```

How about this query ? Would it be pair of arrays ? 

```js
[
  ["?", ".messages[].to", "?to"]
]
```

How do I write policy saying that every message must have some recipient at `*@web.mai` ? Perhaps something like this

```js
[
  ["?", ".messages[].to", "?to"]
  ["every", ["some", ["match", "?to[]", "*@web.mail"]]
]
```

Reply:
> Wait, back up: don't we already get cardinality from the path selector?
> 
> ```js
> // Matching against: {"foo": {"bar": [1,2,3]}}
> 
> policy: [
>   ["$", ".foo.bar", "?x"],
>   ["==", "?x", [1, 2, 3]], // Single cardinality containing an array
>   // the rest is redundant in this context, but for example:
>   ["?x", ".[]", "?y"], // Bind all of x's elements to y
>   ["some", "?y"],
>   ["==", "?y", 2] // matches the `2` in the array
> ]
> ```
> 
> Does that give us what we need without special operators, or have I missed something? 🤔

Ok so let's assume `x` is bound to `[1, 2, 3]`, how is this going to unify ` ["?x", ".[]", "?y"]` ? Since `1 != 2 != 3` unification will fail would not it ? For the last of the `["==", "?y", 2]` to pass `y` should either be bound to 1, 2, and 3 at the same time or each substitution candidate would have to be considered separately somehow and one where it's bound to `2` would need to succeed.

Now that I think of it perhaps something like this could be made to work

```js 
[
  ["?", ".foo.bar", "?x"],
  ["==", "?x", [1, 2, 3]], // Single cardinality containing an array
  // the rest is redundant in this context, but for example:
  ["some", "?x", "?y"], // Bind y to union of all x's elements
  ["==", "?y", 2] // matches `2` because union would match 1 and 3 also
]
```

Not sure if creating special union binding could lead to problems in more complex queries though where in may get unified another union binding.

Reply:
I'm still not sure how one would go about doing something like this

```js
[
  ["?", ".foo.bar", "?x"],
  ["==", "?x", [1, 2, 3]], // Single cardinality containing an array
  // the rest is redundant in this context, but for example:
  ["every", "?x", "?y"], // bind y to all elements at the same time ?
  [">", "?y", 0] // ??
]
```

I suppose y could be bound to some special `Every` binding and on unification it would have to unify with each element or fail. All operators would also need to know to apply to every element I think.

Reply:
> Is the idea that when you omit the first variable, it's assumed to act on args? I can work with that, though having form consistency is also a nice property

no idea was to move selector and all other operators / relations into first position. Entity moved into second and became optional in selectors.

We could keep entry as required in selectors, but changing order seems more consistent imo

Reply:
> This is confusing to me, is it does not map to notation used by datomic.

It should be 😅 I missed a step:

```js
 "cond": [
     ["$args", "foo..bar.[].baz", "?x"],
     ["key", "?x", "?key"], // Should have this
     ["match", "?key", "*@example.com"],
     ["?x", "quux.[3]", 42]
  ]
```

Reply:
> What does [some ?x] imply that x got bound ?

It just means "the resulting set of values of this relation is nonempty"

Reply:
> What should be the ?to bound to in this query ? Array of all emails addresses or each email address ?
> 
> [
>   ["?", ".messages[].to[]", "?to"]
> ]

I would write this

```js
[
  "$args", ".messages.[].to.[]"`, "?to"]
]
```

> How about this query ? Would it be pair of arrays ?
>
> [
>   ["?", ".messages[].to", "?to"]
> ]

Yes, I would expect this to result in a single joined array of arrays:

```
[["alice@web.mail", "malory@web.mail"], ["bob@gmail.com", "alice@web.mail"]]
```

Which you can then select into:

```js
[                     // vvv
  ["$", ".messages.[].to.[1]", "?firstemail"],
  ["==", "?firstemail", "alice@web.mail"]
]
```

Reply:
In other words, I think paths performs an joins on any matching results.

Reply:
> Ok so let's assume x is bound to [1, 2, 3], how is this going to unify ["?x", ".[]", "?y"] ? Since 1 != 2 != 3 unification will fail would not it ? For the last of the ["==", "?y", 2] to pass y should either be bound to 1, 2, and 3 at the same time or each substitution candidate would have to be considered separately somehow and one where it's bound to 2 would need to succeed.

In interesting; that's not how I read the original at all. I think perhaps your approach unifies much earlier in the pipeline than I would expect?

I think of this as a selector that tries to find elements that match the relation (i.e. relational programming). That means that `?x` isn't bound to a single element — it's always bound to the set of all possible matches at that point in the pipeline. You basically take the cross product, and see if your match is anywhere in there. (That's at least how most Datalogs, Kanrens, and SQL work, but maybe Datomic does things differently?)

Reply:
> I suppose y could be bound to some special Every binding and on unification it would have to unify with each element or fail. All operators would also need to know to apply to every element I think.

I don't think it's the operators, it's the pipeline that has to know this. You can always reframe relations to operating on streams of results  (they're isomorphic).

The way that I'm currently thinking about this is that your path selector emits values, and if you are emitting a stream of objects, you need to be able to unify (filter) on each of those objects. You can also do things like filter the keys, and continue further selecting down on a new variable.

To reuse the earlier example:

```js
"cond": [
  // Initial path
  ["$args", "foo..bar.[].baz", "?x"],

  // Bind / unify ?x's keys to ?key
  ["key", "?x", "?key"],

  // on each key, run this filter
  // Due to unification, this reduces the number of ?x values
  ["match", "?key", "*@example.com"],

  // Continue pathing on the reduced set of ?x values
  ["?x", "quux.[3]", 42]
]
```

(Remember that statement order doesn't matter; I've just ordered it this way 
for explanation as a pipeline)

Reply:
> notation used by datomic

Also, perhaps worth calling out: I was also getting confused yesterday about how datalogs work, but this is not a datalog. Datalog is really meant as a database _selection_ language. Kanren is much broader in use case. They're both relational programming languages. (I actually think Datomic has borrow some stuff from Kanren, so there's a fuzzy line here).

I think we're actually more in the Kanren design space than capital-D "Datalog" due to the domain of syntax validation.

Reply:
Also just as an experiment adapting the earlier example with a non-relational version based on FP:

```js
cond: [
  ["pipe", ["$args", "foo..bar.[].baz.[]"],
     [
       ["filter", "x" ["match", ["key", "x"], "*@example.com"],
       ["quux.[3]"],
       ["filter", "x", ["==", 42]]
     ]
  ]
]
```

Which is roughly:

```elixir
get(args, ".foo..bar.[].baz.[]")
  |> filter(fn x -> match (key(x)), "*@example.com" end),
  |> get("quux.[3]")
  |> filter(fn x -> x == 42 end))
]
```

Reply:
(I'm just running an errand; I'll be back shortly!)

Reply:
Oh also: I ran this syntax x past some normie JS devs earlier today (my partner's sister + friends who are mostly FE), and they were skeptical when I told them that we had defined a relational DSL, but were actually *very* positive when they saw examples syntax. Easy to read, way easier to express than e.g. AWS Cedar, and so on

Reply:
I think I need to see an example with `every` because some is kind of default behavior.

> ["$", ".messages.[].to.[]", "?email"], // All elements because that []
> ["some", "?email"], // Must unify != []
> ["==", "?email", "alice@web.mail"] // Filter by alice's email

I find it confusing personally. ?email go bound to array then last clause compares it to string presumably it would filter out non matches ? 

What if operand of some was not resolved to an array ? 

I’m pretty sure it’s not the way datomic, minikarnen nor sicp logic programming work. I re-read those last night. You can unify lists but they just unify elements by position

Reply:
> ?email go bound to array

What I'm trying to express is that `?email` binds _each_ element of the array, not the array itself.

> I think I need to see an example with every because some is kind of default behaviour.

I propose (elsewhere) that the top variables are implicitly bound to `every`.

```js
["every", "?email"], // Implied

["$", ".messages.[].to.[]", "?email"],
["some", "?email"],
["==", "?email", "alice@web.mail"]
```

Reply:
> You can unify lists but they just unify elements by position

Hmm 🤔 

They do have a concept of membership (e.g. `membero`), but you're right that it's slightly different from here.

Raw Datalog (as opposed to Kanren) doesn't have a concept of array position (unless you express it as a linked list or have a special operator). You never really get a single value — it's always matching on the world that it knows about unless you specifically ask to negate all values but one.

Really what I'm trying to do with `.[]` is treat it as a "for each value" rather than matching on the array itself (which is how the path selection works — it treats `.foo.[].bar.baz` as mapping `.bar.baz` over all elements of `.foo`).

I think I see what you're trying to do though: make it referentially transparent, which is a virtue for sure.

Reply:
How would you express "all of the emails in the array at `foo.bar` match *exmaple.com"? And same for "at least one must" and "none must".

Reply:
> It's like "some", but not "every"

In classical datalog, you express every along these lines:

```js
// Triples
[obj, attr-foo, ?x]
[?x, attr-bar, 42]

[obj, attr-foo, ?y]
[eq, ?x, ?y]
```

I think we can do roughly the same in Kanren, no? And then include a relation that does the second variable bindng & comparison under the hood

```js
// Triples
[every, ?x, ?y]
[obj, attr-foo, ?x]
[?y, attr-bar, 42]
```

Reply:
> What I'm trying to express is that ?email binds each element of the array, not the array itself.

Ah I see now that selector gets all recipients. I think I wan to see example where what you select is arrays and not all elements across arrays.

 > I propose (elsewhere) that the top variables are implicitly bound to every unless they're otherwise bound to some

This is confusing to me I would expect the opposite. But even so, would not the query above than fail then because dataset is like on shown below ?

---

I think having concrete examples would really help, here. Assuming following args

```js
{
   "messages": [
      { "to": ["alice@web.mail", "malory@web.mail"], "body": "hello alice" },
      { "to": ["bob@gmail.com", "alice@web.mail"], "body": "hi bob" }
   }
}
```

Can you please show the what would following policies look like ?

1. Every message MUST must have some recipient that matches `*@web.mail`.
2. Some message MUST have all recipients that matches `*@web.mail`.
3. Some message MUST have some recipient that matches `*@web.mail` and a recipient that does not match `*@web.mail`.

Reply:
On my errand, I thought more about the non-relational version of these, too:

```
cond: [
  ["every", "$.foo.bar.[3].baz", ["match", "*.example.com"]]
]
```

Which is a little Haskell-y. It isn't expressive enough to say "the 3rd element of `.foo` must equal the 10th element of `.bar`", but it may be expressive _enough_ for our purposes.

I initially reached for relational programming because 1. it's more than expressive for the things we want to do, and 2. it would get us recursive descent and filtering intermediate parts of the path. We pretty much have that now I think, thanks to agreeing on the selector semantics (we need to settle on the exact syntax for `.foo.[]` vs `.foo[]` and `?` vs `$`, but otherwise I think it's pretty good).

If we use function pipelining, we're more restricted than relations, but it may be large enough?

```js
/// Relational
["$", "foo..bar.[].baz", "?x"], // Start selecting
["key", "?x", "?key"],
["match", "?key", "*@example.com"], // Filter
["?x", "quux.[3]", 42] // Continue to equality match

// Pipelines
["$.foo..bar.[].baz",
  ["filter", ["each-key", ["match", "*@example.com"],
    ["quux.[3]", ["==", 42]]
  ],
]
```

It's less elegant, but it's very clear that you're mapping over elements, and how many you care about. It's a lot closer to your original version, too

Reply:
> I think having concrete examples would really help, here

Agreed 👍 

```
{ 
  "messages": [ 
    { "to": ["alice@web.mail", "malory@web.mail"], "body": "hello alice" }, 
    { "to": ["bob@gmail.com", "alice@web.mail"], "body": "hi bob" } 
  ]
}
```

> Every message MUST must have some recipient that matches *@web.mail.

```js
[
  ["$", ".messages.[].to.[]`, "?tos"],
  ["some", "?tos"],
  ["match", "?tos", "*@web.mail"]
]
```

> Some message MUST have all recipients that matches *@web.mail.

```js
[
  ["$", ".messages.[]`, "?messages"],
  ["some", "?messages"],
  ["?messages", ".to.[]", "?tos"],
  ["match", "?tos", "*@web.mail"]
]
```

> Some message MUST have some recipient that matches *@web.mail and a recipient that does not match *@web.mail.

```js
[
  ["$", ".messages.[]`, "?messages"],
  ["some", "?messages"],

  ["?messages", ".to.[]", "?with-tos"],
  ["some", "?with-tos"],
  ["match", "?with-tos", "*@web.mail"]

  ["?messages", ".to.[]", "?without-tos"],
  ["some", "?with-tos"],
  ["not", ["match", "?with-tos", "*@web.mail"]]
]
```

Reply:
I just flipped back through the relational chapter of SICP (I really do need to give it a full read), but AFAICT they only go as far as a query language. You can extend this (like Kanren (or Souffle on the Datalog-side)) but our use case does require a little bit more than what they provide.

We do care about the number of results that come back from a filtered query (none, some/one-or-more, every/all), and need to fail validation if they return the wrong number of results. This can be expressed directly via unification, but it's very clunky and you need to know how to interpret the results, so UCAN needs to know how to do that at the top level.

---

Frames: SICP gives the streaming form of a relational engine (which is great 👍). To the earlier conversation, the way that I've been thinking about the path selectors (which are effectively sugar for nested unification on keys) is that they basically modify the frames coming in. Our frames are not all of the same size or shape: they can be arbitrary IPLD including nested data, so a path selector can unpack an array or object and push the (flattened) individual elements into the frame stream.

For example:

```js
{
  foo: [1,2,3]
}
```

* `.foo` -> push `[1,2,3]` into the stream
  * Frames: `[..., [1,2,3], ...]`
* `.foo.[]` -> push each element `1`, `2`, and `3` into the stream
  * Frames: `[..., 1, 2, 3, ...]`

It's entirely possible that I'm thinking about this wrong, but that's how I was thinking about how to not need separate operators for collections

Reply:
> [
>  ["$", ".messages.[].to.[]", "?tos"],
>  ["some", "?tos"],
>  ["match", "?tos", "*@web.mail"]
> ]

I would expect this to still apply if one message had matching and other didn’t. Am I missing something ?

Reply:
You know, if it's helpful, we could probably get Will Byrd's take on the problem space

Reply:
> I would expect this to still apply if one message had matching and other didn’t. Am I missing something ?

Maybe? Perhaps mixing querying and validation is confusing 🤔 

Would this more verbose version look closer to what you'd expect?

```js
[
  ["$", ".messages.[].to", "?to"],
  ["every", "?to", "?to-check"],
  ["?to-check", ".[]", "?tos"],
  ["some", "?tos"],
  ["match", "?tos", "*@web.mail"]
]
```

Reply:
(Update in case you see in your email: I managed to make it more concise above)

Reply:
oops, sorry let me try that again:

```js
[
  ["$", ".messages.[], "?msg"],
  ["every", "?msg", "?msg-check"],
  ["?msg-check", ".to.[]", "?tos"]
  ["some", "?tos"],
  ["match", "?tos", "*@web.mail"]
]
```

Reply:
I think that the strength of this approach, though clearly you have to think about _how_ you're recursing, is that you can be clear about how each level behaves. From the above:

```js
[
  // Initial scope of individual frames...
  ["$", ".messages.[]", "?msg"],

    // ...all of which must...
    ["every", "?msg", "?every-msg"],

      // ..have a to field where...
      ["?every-msg", ".to.[]", "?tos"]

        // ...some elements...
        ["some", "?tos"],

        // ...have entries that match @web.mail
        ["match", "?tos", "*@web.mail"]
]

Reply:
Just kind of an observation:

* Finding `some` or `none` is fairly trivial with a relational engine
  * Run the query
  * Were `some` (more than one) or `none` (exactly zero) returned

`every` requires either knowing the size of the starting set, or negating all of your clauses and checking that it's empty. That reads poorly for a common case.

Also note that all of `none`/`some`/`every` are all forms of (stratified) aggregation, and cause special behaviour: success/failure instead of selection.

I've scheduled some time with Quinn tomorrow to see if she has any insight into making this as consistent as possible.

Reply:
I don't want to do this syntax, but parroting (e.g.) Datomic here:

```js
{
  every: ["?x", "?y"], // A little bit like select ... in
  some: ["?z"],
  in: [
    ["$", ".foo.bar.[]", "?x"],
    // ...
  ]
}
```

Reply:
I think we don't need anything for `some` or `none` in minkernen / sicp if it's none it does not match, some that match are the match. There is no every which is also probably why it's difficult to express in syntax expired from them.

I believe you are correct that `every` is an aggregation and those engines don't have it AFIK. That said I think `every` could be done similar to how `and` works in those engines (through recursion), but I need to try it to get some confidence.

How would you feel about leaving out `every` for now and we can consider adding it in the future or maybe even in 1.0 if we have a simple enough implementation strategy. I'm just afraid of making it too difficult for someone to implement it.

Reply:
> if it's none it does not match, some that match are the match.

Ah! This may be the confusion: I'm not thinking of them as constraints. I'm thinking of them as aggregate matchers. This is the part that actually drives the validation, as opposed to selection.

> I think every could be done similar to how and works in those engines (through recursion), but I need to try it to get some confidence.

I'd have to see how. I think that we're mixing two domains: querying and validation. We don't want to see what's in the database: we want to make sure that every record behaves a certain way (more like a database index constraint than a query).

> How would you feel about leaving out every for now

I literally don't think we can, unless I'm missing some major detail.

Reply:
As a concrete example, if you run this _query_:

```js
{
  policy: [
    ["$", ".foo.bar.[]", "?x"],
    ["match", "?x", "*@example.com"]
  ]
}
```

What do you expect the behaviour to be? In a query engine, if this returns no results, that's fine. I guess it's `some` by default? If so, I think we get unexpected behaviour for this:

```js
{
  policy: [
    ["$", ".foo.bar.[]", "?x"],
    [">", "?x", 42]
  ]
}
```

This ☝️ would be "some of the values need to be bigger than 42", which is not what I would probably want to do with numeric comparison most of the time.

## Empty Count

If we include a `count` aggregate, I think these problems go away (and there's a way to express `every` without a special aggregate)

```js
{
  policy: [
    ["$", ".foo.bar.[]", "?x"],
    [">", "?x", 42]
    ["not", ["==", ["count", "?x], 0] // AKA `[not [empty ?x]]`
  ]
}
```

How does that feel?

Reply:
I tried a few things that use our implicit top-level `count` (to see that `some` records have been returned), but without access to the keyword I don't think it's possible.

To your earlier point: can we move forward with `some` for now and add `every` at some point in the near-to-medium term? Sure, I do need to ship this library, so I'm on board. I have a meeting with Quinn (and maybe Philipp) tomorrow about these topics so I'll report back about that, too. But yes: shipping is good, as long as we don't close the door on features that I know I have use cases for.

Reply:
> I think we don't need anything for some or none in minkernen / sicp

Wait, how are you thinking about expressing `none` without aggregation?

Reply:
Just stashing some research here: this paper adds universal quantification to Kanren: https://www.cs.toronto.edu/~lczhang/jin_universal2021.pdf

It adds a form `(forall (v) domain goal)`

Translated to our format it looks roughly like this:

[Note: adopting polish notation]

```js
{ 
  policy: [ 
    ["$", ".foo.bar.[]", "?x"],
    ["forall", "?x", "?y",]
    ["match", "?y", "*@example.com"]
  ]
}
```

...which looks a heck of a lot like my `every` suggestion from earlier.

What I like about this is we could add this to the version with an implied `some` without having to rework the version that doesn't have it 👍

Reply:
That paper makes a ton of sense (though I think we probably need to avoid `implies` for security reasons). Proposed _future_ syntax for that `forall` (I emphasise: _future_, because we should just ship `some` first).

```js
{ 
  policy: [ 
    // Every item at $.foo.bar.[] must contain a sub-DAG
    // at .baz.[] with at least one value that matches *@example.com

    [".foo.bar.[]", "$", "?domain"],
    ["every", ["?goal"], "?domain"], // "every ?goal in ?domain"

    [".baz.[]", "?goal", "?inner"]
    ["match", "?inner", "*@example.com"]
  ] 
}
```

Reply:
Ah, also a simplification for the engine: we don't need to do any graph analysis to know which variables are (implied) to be what we're looking for. We only need to host variables of the form `[<path>, "$", "?goal"]`

Reply:
Did a second deep read of the paper. They have to handle much more general cases than us (arbitrary domains, user-defined relations, etc).

`forall` is all-or-nothing, so we could probably just return bottom as soon as the first predicate fails, and propagate up. `forall`/`every` does need that second logic variable, but that's unavoidable.

I don't _think_ that we need to stratify foralls because their behaviour is monotone (if they fail, they fail forever).

Reply:
Forked this into https://github.com/ucan-wg/delegation/issues/8